File: Chapter_01/binary_search.cpp
#include <iostream>
#include <vector>

int binary_search(const std::vector<int> &arr, int item);

int main()
{
    std::vector<int> data = {1, 2, 3, 4, 5};
    std::cout << binary_search(data, 2) << "\n";
    std::cout << binary_search(data, 5) << "\n";
    std::cout << binary_search(data, -2) << "\n";

    return 0;
}

int binary_search(const std::vector<int> &arr, int item)
{
    int low = 0;
    int high = arr.size() - 1;
    int mid;
    int guess;
    while (low <= high)
    {
        mid = (low + high) / 2;
        guess = arr[mid];
        if (item == guess)
        {
            return mid;
        }
        else if (item < guess)
        {
            high = mid - 1;
        }
        else
        {
            low = mid + 1;
        }
    }
    return -1;
}

File: Chapter_01/binary_search.py
#!/usr/bin/python3

def binary_search(arr, item):
    low = 0
    high = len(arr)-1
    while low <= high:
        mid = (low + high) // 2
        guess = arr[mid]
        if item == guess:
            return mid
        elif item < guess:
            high = mid - 1
        else:
            low = mid+1
    return None


my_list = [1, 2, 3, 4, 5, 6]

print(binary_search(my_list, 3))
print(binary_search(my_list, -1))
